tidy3d/plugins/microwave/custom_path_integrals.py

```diff
@@ -0,0 +1,163 @@
+        2, title="Axis", description="Specifies dimension of the planar axis (0,1,2) -> (x,y,z)."
+    )
+
+    position: float = pd.Field(
+        ...,
+        title="Position",
+        description="Position of the plane along the ``axis``.",
+    )
+
+    vertices: ArrayFloat2D = pd.Field(
+        ...,
+        title="Vertices",
+        description="List of (d1, d2) defining the 2 dimensional positions of the path. "
+        "Dimensions should be ordered to maintain a right-handed coordinate system. "
+        "If the axis corresponds with ``y``, vertices should correspond with ``(z, x)``. "
+        "If you wish to indicate a closed contour, the final vertex should be made "
+        "equal to the first vertex. ``vertices[-1]==vertices[0]``",
+        units=MICROMETER,
+    )
+
+    def compute_integral(
+        self, field: FieldParameter, em_field: MonitorDataTypes
+    ) -> IntegralResultTypes:
+        """Computes the path integral defined by ``vertices`` given the input ``em_field``.
+        The ``FieldParameter`` determines whether the electric or magnetic field is integrated."""
+        if not isinstance(em_field, (FieldData, FieldTimeData, ModeSolverData)):
+            raise DataError("'em_field' type not supported.")
+
+        dim1 = "xyz"[self.remaining_axes[0]]
+        dim2 = "xyz"[self.remaining_axes[1]]
+        dim3 = "xyz"[self.main_axis]
+
+        h_field_name = f"{field}{dim1}"
+        v_field_name = f"{field}{dim2}"
+        # Validate that fields are present
+        if h_field_name not in em_field.field_components:
+            raise DataError(f"'field_name' '{h_field_name}' not found.")
+        if v_field_name not in em_field.field_components:
+            raise DataError(f"'field_name' '{v_field_name}' not found.")
+
+        # Select fields lying on the plane
+        plane_indexer = {dim3: self.position}
+        field1 = em_field.field_components[h_field_name].sel(plane_indexer, method="nearest")
+        field2 = em_field.field_components[v_field_name].sel(plane_indexer, method="nearest")
+
+        # Dim 's' represents the parameterization of the line
+        # 't' is likely used for time
+        x_path = xr.DataArray(self.vertices[:, 0], dims="s")
+        y_path = xr.DataArray(self.vertices[:, 1], dims="s")
+        field1_interp = field1.interp(x=x_path, y=y_path, method="linear")
+        field2_interp = field2.interp(x=x_path, y=y_path, method="linear")
+
+        # Determine the differential length elements along the path
+        dl_x = self._compute_dl_component(x_path, self.is_closed_contour)
+        dl_y = self._compute_dl_component(y_path, self.is_closed_contour)
+        dl_x = xr.DataArray(dl_x, dims="s")
+        dl_y = xr.DataArray(dl_y, dims="s")
+
+        # Compute the dot product between differential length element and vector field
+        integrand = field1_interp * dl_x + field2_interp * dl_y
+        # Integrate along the path
+        result = integrand.integrate(coord="s")
```

Comment:
nice 👍 so this DataArray indexing seems to work pretty well?

Reply:
Yea, its pretty simple!